# Assignment 2: Linear Models and Validation Metrics (30 marks total)
### Due: October 10 at 11:59pm

### Name: 

### In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.

## Part 1: Classification (14.5 marks total)

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:

### Step 0: Import Libraries

In [62]:
import numpy as np
import pandas as pd

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [63]:
# TO DO: Import spam dataset from yellowbrick library

from yellowbrick.datasets import load_spam
X, y = load_spam()
# TO DO: Print size and type of X and y
print("Size of X: " , X.shape)
print("Type of X: ", type(X))
print("Size of y: " , y.shape)
print("Type of y: ", type(y))


Size of X:  (4600, 57)
Type of X:  <class 'pandas.core.frame.DataFrame'>
Size of y:  (4600,)
Type of y:  <class 'pandas.core.series.Series'>


### Step 2: Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [64]:
# TO DO: Check if there are any missing values and fill them in if necessary
anyMissingValue = X.isnull().sum()
print(anyMissingValue)



word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_our                 0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq_will                0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet              0
word_fre

For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [65]:
# TO DO: Create X_small and y_small 
from sklearn.model_selection import train_test_split
X_small_train, X_small_test, y_small_train, y_small_test  = train_test_split(X, y, random_state = 0, train_size = 0.05) 

### Step 3: Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets: 
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

In [66]:
# 3.1 Importing LogisticRegression from sklearn
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

# 3.2 Instantiate model LogisticRegression(max_iter = 2000)
logreg = LogisticRegression(max_iter=2000)

# 3.3 Implement the machine learning model with three different datasets
# a) set for  X and y
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0)

# b) For only two columns of X and y
X_subset = X.iloc[:, :2]
X_train_selective, X_test_selective, y_train_selective, y_test_selective = train_test_split(X_subset, y, random_state = 0)

# c) For  X_small and y_small

X_small_train, X_small_test, y_small_train, y_small_test  = train_test_split(X, y, random_state = 0, train_size = 0.05) 






### Step 4: Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

In [67]:

# Validation for the first set(compkete Data Set)
logreg1 = LogisticRegression(max_iter=2000).fit(X_train, y_train)
print("Training set score for complete set: {:.3f}".format(logreg1.score(X_train, y_train)))
print("Validation set score for complete set: {:.3f}".format(logreg1.score(X_test, y_test)))
print('')


# Validation for the second set(Data set that contains First two columns)
logreg2 = LogisticRegression(max_iter=2000).fit(X_train_selective, y_train_selective)
print("Training set score for set(contains first two columns): {:.3f}".format(logreg2.score(X_train_selective, y_train_selective)))
print("Validation set score for set(contains first two columns): {:.3f}".format(logreg2.score(X_test_selective, y_test_selective)))
print('')


# Validation for the first set()(Data set that contains 5% of the data)
logreg3 = LogisticRegression(max_iter=2000).fit(X_small_train, y_small_train)
print("Training set score for set(that contains 5 percentage of the data): {:.3f}".format(logreg3.score(X_small_train, y_small_train)))
print("Validation set score for set(that contains 5 percentage of the data): {:.3f}".format(logreg3.score(X_small_test, y_small_test)))










Training set score for complete set: 0.929
Validation set score for complete set: 0.938

Training set score for set(contains first two columns): 0.608
Validation set score for set(contains first two columns): 0.613

Training set score for set(that contains 5 percentage of the data): 0.948
Validation set score for set(that contains 5 percentage of the data): 0.907


### Step 5: Visualize Results (4 marks)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [68]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT

from sklearn.linear_model import LogisticRegression
data = {
  'Data size' : [len(X_train),len(X_train_selective),len(X_small_train)],
  'training accuracy' : [logreg1.score(X_train, y_train),logreg2.score(X_train_selective, y_train_selective), logreg3.score(X_small_train, y_small_train)],
  'validation accuracy' : [logreg1.score(X_test, y_test), logreg2.score(X_test_selective, y_test_selective), logreg3.score(X_small_test, y_small_test)  ]
}
results = pd.DataFrame(data)
print(results)

   Data size  training accuracy  validation accuracy
0       3450           0.928696             0.938261
1       3450           0.608406             0.613043
2        230           0.947826             0.906636


### Questions (4 marks)
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

*YOUR ANSWERS HERE*
Answer 1- Full Data Set (3450 samples):
Training Accuracy: 92.87%
Validation Accuracy: 93.83%

First Two Columns of Data (3450 samples):
Training Accuracy: 60.84%
Validation Accuracy: 61.30%

Only 5% of the Data (230 samples):
Training Accuracy: 94.78%
Validation Accuracy: 90.66%

I observed that , Training accuracy tends to remain high even with a smaller data size(comparing 92.87% with full data vs 94.78% with 5% of data), indicating that the model can fit the available data well. However, Validation accuracy may slightly decrease (from 93.83% to 90.66%) with smaller data sizes but can still be relatively high if the subset used for validation is representative.
On the other hand, using a limited subset of features (first two columns) can lead to a significant drop in accuracy(Training Accuracy: 60.84% and Validation Accuracy: 61.30% ), suggesting the importance of feature selection.


Answer 2- False Positive : A false positive occurs when the spam filter incorrectly classifies a legitimate email as spam. 
False Negative : A false negative occurs when the spam filter incorrectly classifies a spam email as legitimate
In my opinion, the consequence of a false negative is potentially worse. It means that unwanted spam emails are not filtered out and may clutter the user's inbox. Some of these spam emails could be phishing attempts or contain malware, posing security risks to the user.


### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*
Q1 Where did you source your code?
Answer:- The main source of the code was the practice example given by the instructor in D2L. In addition, I used ChatGPT to learn some of the instructions.

Q2 In what order did you complete the steps?
Answer :- 1. Loading and arrange data into feature matrix and target vector
2. Checking any missing values
3. Choose model class
4. Instantiating model and selecting hyperparameters
5.Fitting model to data
6.Predict values for new data
7. Comparing accuracy.

Q3 If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
Answer: - I used some of the prompts like:-
a. 	What are touple object and how to check it, can you give me an example of touple in dataset?
b. 	How to have subset of dataset?  And I modified the code accordingly.
c. 	How to make a loop to insert the data values in the frame
I had to make some changes to my code as I was not able to take subset of data.

Q4 Did you have any challenges? If yes, what were they? If not, what helped you to be successful?
Answer- Yes, I faced some difficulties like I was not able to split the data or how the regression models works. Then, I took the help of lecture notes and examples on D2L.


## Part 2: Regression (10.5 marks total)

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [46]:
# TO DO: Import spam dataset from yellowbrick library
from yellowbrick.datasets import load_concrete
X,y = load_concrete()
# TO DO: Print size and type of X and y
print("Size of X: " , X.shape)
print("Type of X: ", type(X))
print("Size of y: " , y.shape)
print("Type of y: ", type(y))




Size of X:  (1030, 8)
Type of X:  <class 'pandas.core.frame.DataFrame'>
Size of y:  (1030,)
Type of y:  <class 'pandas.core.series.Series'>


### Step 2: Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [47]:
# TO DO: Check if there are any missing values and fill them in if necessary
anyMissingValue = X.isnull().sum()
print(anyMissingValue)

cement    0
slag      0
ash       0
water     0
splast    0
coarse    0
fine      0
age       0
dtype: int64


### Step 3: Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Implement the machine learning model with `X` and `y`

In [55]:
# TO DO: ADD YOUR CODE HERE
# Note: for any random state parameters, you can use random_state = 0

# 1. Importing LinearRegression from sklearn
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

# 2.  Instantiate model LogisticRegression(max_iter = 2000)
lr = LinearRegression()

# 3. Implement the machine learning model with X and y
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0)

lr.fit(X_train, y_train)

LinearRegression()

### Step 4: Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [57]:
# TO DO: ADD YOUR CODE HERE
print("Mean Squared Training score: {:.2f}".format(mean_squared_error(y_train, lr.predict(X_train))))
print("Mean Squared Validation score: {:.2f}".format(mean_squared_error(y_test, lr.predict(X_test))))
print('')
print("R2 Training score: {:.2f}".format(r2_score(y_train, lr.predict(X_train))))
print("R2 Valdiation score: {:.2f}".format(r2_score(y_test, lr.predict(X_test))))

Mean Squared Training score: 111.36
Mean Squared Validation score: 95.90

R2 Training score: 0.61
R2 Valdiation score: 0.62


### Step 5: Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [60]:
# TO DO: ADD YOUR CODE HERE
data = {
  'Training Accuracy' : [round(mean_squared_error(y_train, lr.predict(X_train)), 2), round(r2_score(y_train, lr.predict(X_train)), 2)],
  'Validation Accuracy' : [round(mean_squared_error(y_test, lr.predict(X_test)), 2), round(r2_score(y_test, lr.predict(X_test)), 2) ]
}
results = pd.DataFrame(data, index = ['MSE', 'R2 score'])
print(results)


          Training Accuracy  Validation Accuracy
MSE                  111.36                95.90
R2 score               0.61                 0.62


### Questions (2 marks)
1. Did using a linear model produce good results for this dataset? Why or why not?
Answer:- In this case, the linear model didn't perform too well. The Mean Squared Error (MSE) for training was about 111.36, and for validation, it was around 95.90 . The R2 scores, which show how well the model explains the data, were only about 0.61 for training and 0.62 for validation.
These numbers suggest that the linear model struggled to make accurate predictions. The high MSE means the model's predictions were quite off, and the R2 scores indicate that it could only explain around 60% of the data's variance.
So, using a linear model didn't produce good results for this dataset. It might be better to try other models or approaches to improve accuracy

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*
Q1.  Where did you source your code?
Answer:- The main source of the code was the practice example given by the instructor in D2L. In addition, I used ChatGPT to learn some of the instructions.

Q2.  In what order did you complete the steps?
Answer :- 1. Loading and arrange data into feature matrix and target vector
2. Checking any missing values
3. Choose model class
4. Instantiating model and selecting hyperparameters
5.Fitting model to data
6.Predict values for new data
7. Comparing accuracy.

Q3.  If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
Answer: - I used some of the prompts like:-What are mean square error and R2 Score and how with these parameters define the model
Q4.  Did you have any challenges? If yes, what were they? If not, what helped you to be successful?
Answer- Yes, I faced some difficulties like I was not able to understand R2 score then I took help from google as well..


## Part 3: Observations/Interpretation (3 marks)

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


*ADD YOUR FINDINGS HERE*
Part 1- Full Data Set vs. First Two Columns:
Using the full data set yields high training and validation accuracy, indicating good performance.Using only the first two columns significantly drops accuracy, suggesting the need for more data or informative features.
Full Data Set vs. 5% of the Data:
Even with only 5% of the data, training accuracy remains high.Validation accuracy, though slightly lower, is still good, implying that this small subset may contain critical patterns.
Part 2- "When analyzing the results, a noticeable pattern emerges. The Mean Squared Error (MSE) values for both training and validation data are relatively high, with training at around 111.36 and validation around 95.90. Additionally, the R2 scores, which indicate how well the model fits the data, are approximately 0.61 for training and 0.62 for validation.These findings are consistent with what we discussed during lectures. A high MSE suggests that the linear model struggled to make accurate predictions, and the R2 scores, though not extremely low, indicate that the model could only explain around 60% of the data's variance.


## Part 4: Reflection (2 marks)
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


*ADD YOUR THOUGHTS HERE*
I liked the way this assignment is designed. While making this assignment , I revised all the components that we have learnt. The challanging thing was to cover the MSE and R2 score part

## Part 5: Bonus Question (4 marks)

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [89]:
# TO DO: ADD YOUR CODE HERE
# 1 ) With Ridge regression 
# 1.a. Importing RidgeRegression from sklearn
from sklearn.linear_model import Ridge
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

# 1.b.  Instantiate model Ridge model with an appropriate alpha value
ridge = Ridge(alpha=1)   # alpha value from 0.001 to 37 gives the same value after that it start decreasing 

# 1.c. Implement the machine learning model with X and y
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0)

ridge.fit(X_train, y_train)

print("R2 Training score for Ridge model: {:.2f}".format(r2_score(y_train, ridge.predict(X_train))))
print("R2 Valdiation score for Ridge model: {:.2f}".format(r2_score(y_test, ridge.predict(X_test))))

# 2 ) With Lasso regression 
# 2.a. Importing LassoRegression from sklearn
from sklearn.linear_model import Lasso

# 2.b.  Instantiate model Lasso model with an appropriate alpha value
lasso = Lasso(alpha=0.001)    

# 2.c. Implement the machine learning model with X and y
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0)

lasso.fit(X_train, y_train)

print("R2 Training score for Lasso model: {:.2f}".format(r2_score(y_train, lasso.predict(X_train))))
print("R2 Valdiation score for Lasso model: {:.2f}".format(r2_score(y_test, lasso.predict(X_test))))




R2 Training score for Ridge model: 0.56
R2 Valdiation score for Ridge model: 0.54
R2 Training score for Lasso model: 0.55
R2 Valdiation score for Lasso model: 0.54


*ANSWER HERE*
In practice, an R^2 score above 0.7 or 0.8 is often considered good, indicating that the model captures a substantial amount of the variance in the data. However the Ridge model(at alpha - from 0.01 to 37 gives its maximum value of around .54) and same for lasso(at alpha = 0.01). However, as compared to Linear Regession, both the model are not giving good output.